core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -6030,34 +6040,45 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
       else {
         ZIO.uninterruptibleMask { restore =>
           val promise = Promise.unsafe.make[Unit, Unit](FiberId.None)(Unsafe.unsafe)
-          val ref     = new java.util.concurrent.atomic.AtomicInteger(0)
-          ZIO.transplant { graft =>
-            ZIO.foreach(as) { a =>
-              graft {
-                restore(ZIO.suspendSucceed(f(a))).foldCauseZIO(
-                  cause => promise.fail(()) *> ZIO.refailCause(cause),
-                  _ =>
-                    if (ref.incrementAndGet == size) {
-                      promise.unsafe.done(ZIO.unit)(Unsafe.unsafe)
-                      ZIO.unit
-                    } else {
-                      ZIO.unit
+          val ref     = new java.util.concurrent.atomic.AtomicInteger(size)
+          ZIO
+            .withFiberRuntime[R, Nothing, Iterable[Fiber.Runtime[E, Unit]]] { (fiber, _) =>
+              ZIO.foreach(as) { a =>
+                restore(f(a))
+                  .foldCauseZIO(
+                    cause =>
+                      ZIO.withFiberRuntime[Any, E, Unit] { (childFiber, _) =>
+                        ZIO.suspendSucceed {
+                          childFiber.transferChildren(fiber.scope)
+                          promise.fail(()) *> ZIO.refailCause(cause)
+                        }
+                      },
+                    _ => {
+                      ZIO.withFiberRuntime[Any, Nothing, Unit] { (childFiber, _) =>
+                        ZIO.succeed {
+                          childFiber.transferChildren(fiber.scope)
+                          if (ref.decrementAndGet() == 0) {
+                            promise.unsafe.done(ZIO.unit)(Unsafe.unsafe)
+                          }
+                        }
+                      }
                     }
-                )
-              }.forkDaemon
+                  )
+                  .forkDaemon
```

Comment:
why fork daemon in the first place? wouldn't a simple fork be a better fit here?  
child fibers inherits the fiber-refs, gets interrupted when this fiber completes/fails/interrupts, no need for any explicit error handling...
* another option is to spawn a single fiber which becomes the parent of all child fibers, this fiber joins the child fibers and is joined by the current fiber (the one executing `foreachParXXX`), this allows simple interruption and guaranteeing interruption of all pending fibers

Reply:
> why fork daemon in the first place? wouldn't a simple fork be a better fit here?

I'm not sure; I was contemplating changing it as well but was too worried it might have some repercussions.

> another option is to spawn a single fiber which becomes the parent of all child fibers

I think that this would semantically be the same as using `fork`, but with the extra overhead of having another layer inbetween. If we're going to go down this path I think going with `join` is better, but I'd rather not do it in this PR as I want to minimize any chances of regression as much as possible